config-bot/main

```diff
@@ -0,0 +1,373 @@
+#!/usr/bin/python3 -u
+
+import os
+import re
+import sys
+import json
+import toml
+import fnmatch
+import aiohttp
+import asyncio
+import argparse
+import tempfile
+import subprocess
+
+
+DEFAULT_CONFIG_FILE_PATH = "/etc/config-bot.toml"
+
+git = None
+
+
+def main():
+    args = parse_args()
+    cfg = load_config(args.config)
+
+    global git
+    git = Git(cfg['git'])
+
+    loop = asyncio.get_event_loop()
+
+    o = cfg.get('sync-build-lockfiles')
+    if o is not None:
+        loop.create_task(sync_build_lockfiles(o))
+
+    o = cfg.get('promote-lockfiles')
+    if o is not None:
+        loop.create_task(promote_lockfiles(o))
+
+    o = cfg.get('propagate-files')
+    if o is not None:
+        loop.create_task(propagate_files(o))
+
+    loop.run_forever()
+    loop.close()
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--config', help="Path to TOML config file",
+                        default=DEFAULT_CONFIG_FILE_PATH)
+    return parser.parse_args()
+
+
+def load_config(fn):
+    with open(fn) as f:
+        return toml.load(f)
+
+
+async def sync_build_lockfiles(cfg):
+    # this is the only mode we support right now
+    assert cfg['trigger']['mode'] == 'periodic'
+    period = period_to_seconds(cfg['trigger']['period'])
+    method = cfg['method']
+
+    base_url = cfg['builds-base-url']
+
+    # {stream -> buildid}
+    last_build_ids = {}
+
+    first = True
+    while True:
+
+        # for easier hacking; just act immediately on first iteration
+        # this allows us to safely use `continue` afterwards
+        if not first:
+            await asyncio.sleep(period)
+        first = False
+
+        for stream in cfg['streams']:
+            builds_json = f'{base_url}/{stream}/builds/builds.json'
+            builds = json.loads(await http_fetch(builds_json))
+            if builds is None:
+                # problem fetching the json; just ignore and we'll retry
```

Comment:
Should we add a retry counter and at least print out (if not directly failing) if an individual streams data cannot be retrieved within x tries? With the current setup it could be easy to set it up with a typo in one of the stream names and have the lockfiles never synced.

Reply:
> With the current setup it could be easy to set it up with a typo in one of the stream names and have the lockfiles never synced.

Yeah, I think metrics/alerts of some kind are next so we can actually monitor if the app is running as intended without doing `oc logs`. Could probably start with a basic webpage and e.g. emailing if there are X errors within T? Would be cool to plug into Prometheus as well. Let's split out that discussion into a separate ticket which encompasses coreos-koji-tagger too?

> Should we add a retry counter and at least print out (if not directly failing)

Right now, we print on every failure in `http_fetch`. Because this is a long-running app, I'm not sure I want to stop retrying without some kind of higher-level monitoring & alert in place first.

Reply:
> Could probably start with a basic webpage and e.g. emailing if there are X errors within T? Would be cool to plug into Prometheus as well. 

Yeah, having it in Prometheus would make the emailing on x errors or webpage very easy.

>  Because this is a long-running app, I'm not sure I want to stop retrying without some kind of higher-level monitoring & alert in place first.

That's fair.

Reply:
OK, split this out in https://github.com/coreos/fedora-coreos-releng-automation/issues/5